ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/Committer.scala

```diff
@@ -62,39 +57,88 @@ private[committer] trait Committer[Submission, PartialResult] {
     }.toMap
 
   /** A committer can `run` a submission and produce a log entry and output states. */
-  @SuppressWarnings(Array("org.wartremover.warts.Return"))
   def run(
       recordTime: Option[Time.Timestamp],
-      submission: Submission,
+      submission: DamlSubmission,
       participantId: ParticipantId,
       inputState: DamlStateMap,
   ): (DamlLogEntry, Map[DamlStateKey, DamlStateValue]) =
     runTimer.time { () =>
       val ctx = new CommitContext {
         override def getRecordTime: Option[Time.Timestamp] = recordTime
+
         override def getParticipantId: ParticipantId = participantId
+
         override def inputsWithFingerprints: DamlStateMapWithFingerprints =
           inputState.map {
             case (key, value) => (key, (value, FingerprintPlaceholder))
           }
       }
-      var cstate = init(ctx, submission)
-      for ((info, step) <- steps) {
-        val result: StepResult[PartialResult] =
-          stepTimers(info).time(() => step(ctx, cstate))
-        result match {
-          case StepContinue(newCState) => cstate = newCState
-          case StepStop(logEntry) =>
-            return logEntry -> ctx.getOutputs.toMap
-        }
+      val logEntry = runSteps(ctx, submission)
+      logEntry -> ctx.getOutputs.toMap
+    }
+
+  def dryRun(
+      submission: DamlSubmission,
+      participantId: ParticipantId,
+      inputState: DamlStateMapWithFingerprints,
+  ): PreexecutionResult = {
+    runTimer.time { () =>
+      // TODO(miklos): Create context for pre-execution here.
+      val commitContext = new CommitContext {
+        override def getRecordTime: Option[Time.Timestamp] = None
+
+        override def getParticipantId: ParticipantId = participantId
+
+        override def inputsWithFingerprints: DamlStateMapWithFingerprints = inputState
       }
-      sys.error(s"Internal error: Committer $committerName did not produce a result!")
+      preexecute(submission, participantId, inputState, commitContext)
+    }
+  }
+
+  private[committer] def preexecute(
+      submission: DamlSubmission,
+      participantId: ParticipantId,
+      inputState: DamlStateMapWithFingerprints,
+      commitContext: CommitContext,
+  ): PreexecutionResult = {
+    val logEntry = runSteps(commitContext, submission)
+    PreexecutionResult(
+      readSet = commitContext.getAccessedInputKeysWithFingerprints.toMap,
+      successfulLogEntry = logEntry,
+      stateUpdates = commitContext.getOutputs.toMap,
+      // TODO(miklos): Take out-of-time-bounds log entry and min/max record time from context.
+      outOfTimeBoundsLogEntry = constructOutOfTimeBoundsLogEntry(commitContext),
+      minimumRecordTime = Timestamp.MinValue,
+      maximumRecordTime = Timestamp.MaxValue,
+      // We assume the time updates must be visible to every participant.
```

Comment:
Why is that? Time updates should be received only by involved participants in case of success and only by the submitting participant in case of out-of-time and conflict.

Reply:
Yes, that's correct for a ledger with sub-transaction privacy. However, we don't have such ledger implementations as part of the DAML SDK.

Reply:
Right, I'd only make it a bit more explicit in the comment, e.g. "This default value only covers use cases that don't need sub-transaction privacy".